// @iiif/base-component/dist-esmodule/index.js
var BaseComponent = /** @class */ (function () {
    function BaseComponent(options) {
        this.options = options;
        this.options.data = Object.assign({}, this.data(), options.data);
    }
    BaseComponent.prototype._init = function () {
        this.el = this.options.target;
        if (!this.el) {
            console.warn('element not found');
            return false;
        }
        this.el.innerHTML = '';
        return true;
    };
    BaseComponent.prototype.data = function () {
        return {};
    };
    BaseComponent.prototype.on = function (name, callback, ctx) {
        var e = this._e || (this._e = {});
        (e[name] || (e[name] = [])).push({
            fn: callback,
            ctx: ctx
        });
    };
    BaseComponent.prototype.fire = function (name) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var data = [].slice.call(arguments, 1);
        var evtArr = ((this._e || (this._e = {}))[name] || []).slice();
        var i = 0;
        var len = evtArr.length;
        for (i; i < len; i++) {
            evtArr[i].fn.apply(evtArr[i].ctx, data);
        }
    };
    BaseComponent.prototype._resize = function () {
    };
    BaseComponent.prototype.set = function (_data) {
    };
    return BaseComponent;
}());
export { BaseComponent };
