// @universalviewer/aleph/dist/collection/functional-components/aframe/TrackballCamera.js
import { h } from "@stencil/core";
export const TrackballCamera = ({ animating, aspect, cb, controlPosition, controlTarget, dampingFactor, enabled, far, fov, graphEnabled, maxDistance, near, panSpeed, rotateSpeed, screenHeight, screenWidth, zoomSpeed }, _children) => (() => {
    if (graphEnabled) {
        return (h("al-a-camera", { fov: fov, near: near, aspect: aspect, "look-controls": "enabled: false", far: far, id: "mainCamera", raycaster: "objects: [data-raycastable]", "al-cursor": "rayOrigin: mouse", "al-trackball-control": `
            screenLeft: ${0};
            screenTop: ${0};
            screenWidth: ${screenWidth};
            screenHeight: ${screenHeight};
            rotateSpeed: ${rotateSpeed};
            zoomSpeed: ${zoomSpeed};
            panSpeed: ${panSpeed};
            maxDistance: ${maxDistance};
            dynamicDampingFactor: ${dampingFactor};
            controlTarget: ${controlTarget};
            controlPosition: ${controlPosition};
            enabled: ${enabled};
            animating: ${animating}
          `, "al-control-lights": true, ref: ref => cb(ref) }));
    }
    else {
        return (h("al-a-camera", { fov: fov, near: near, aspect: aspect, "look-controls": "enabled: false", far: far, id: "mainCamera", "al-trackball-control": `
            screenLeft: ${0};
            screenTop: ${0};
            screenWidth: ${screenWidth};
            screenHeight: ${screenHeight};
            rotateSpeed: ${rotateSpeed};
            zoomSpeed: ${zoomSpeed};
            panSpeed: ${panSpeed};
            maxDistance: ${maxDistance};
            dynamicDampingFactor: ${dampingFactor};
            controlTarget: ${controlTarget};
            controlPosition: ${controlPosition};
            enabled: ${enabled};
            animating: ${animating}
          `, "al-control-lights": true, ref: ref => cb(ref) }));
    }
})();
