// @iiif/iiif-av-component/dist-commonjs/helpers/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IS_DEV = process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test';
var isDev = function () { return IS_DEV || (window && window.__DEBUG_AV_COMPONENT__); };
exports.Logger = {
    enabled: false,
    log: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.enabled && isDev()) {
            console.log.apply(console, args);
        }
    },
    warn: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.enabled && isDev()) {
            console.warn.apply(console, args);
        }
    },
    error: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.enabled && isDev()) {
            console.error.apply(console, args);
        }
    },
    groupCollapsed: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.enabled && isDev()) {
            console.groupCollapsed.apply(console, args);
        }
    },
    group: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.enabled && isDev()) {
            console.group.apply(console, args);
        }
    },
    groupEnd: function () {
        if (this.enabled && isDev()) {
            console.groupEnd();
        }
    },
};
