// @edsilv/key-codes/dist-esmodule/index.js
export var KeyDown;
(function (KeyDown) {
    KeyDown[KeyDown["Backspace"] = 8] = "Backspace";
    KeyDown[KeyDown["Tab"] = 9] = "Tab";
    KeyDown[KeyDown["Enter"] = 13] = "Enter";
    KeyDown[KeyDown["Shift"] = 16] = "Shift";
    KeyDown[KeyDown["Ctrl"] = 17] = "Ctrl";
    KeyDown[KeyDown["Alt"] = 18] = "Alt";
    KeyDown[KeyDown["PauseBreak"] = 19] = "PauseBreak";
    KeyDown[KeyDown["CapsLock"] = 20] = "CapsLock";
    KeyDown[KeyDown["Escape"] = 27] = "Escape";
    KeyDown[KeyDown["Spacebar"] = 32] = "Spacebar";
    KeyDown[KeyDown["PageUp"] = 33] = "PageUp";
    KeyDown[KeyDown["PageDown"] = 34] = "PageDown";
    KeyDown[KeyDown["End"] = 35] = "End";
    KeyDown[KeyDown["Home"] = 36] = "Home";
    KeyDown[KeyDown["LeftArrow"] = 37] = "LeftArrow";
    KeyDown[KeyDown["UpArrow"] = 38] = "UpArrow";
    KeyDown[KeyDown["RightArrow"] = 39] = "RightArrow";
    KeyDown[KeyDown["DownArrow"] = 40] = "DownArrow";
    KeyDown[KeyDown["PrintScrn"] = 44] = "PrintScrn";
    KeyDown[KeyDown["Insert"] = 45] = "Insert";
    KeyDown[KeyDown["Delete"] = 46] = "Delete";
    KeyDown[KeyDown["Zero"] = 48] = "Zero";
    KeyDown[KeyDown["One"] = 49] = "One";
    KeyDown[KeyDown["Two"] = 50] = "Two";
    KeyDown[KeyDown["Three"] = 51] = "Three";
    KeyDown[KeyDown["Four"] = 52] = "Four";
    KeyDown[KeyDown["Five"] = 53] = "Five";
    KeyDown[KeyDown["Six"] = 54] = "Six";
    KeyDown[KeyDown["Seven"] = 55] = "Seven";
    KeyDown[KeyDown["Eight"] = 56] = "Eight";
    KeyDown[KeyDown["Nine"] = 57] = "Nine";
    KeyDown[KeyDown["a"] = 65] = "a";
    KeyDown[KeyDown["b"] = 66] = "b";
    KeyDown[KeyDown["c"] = 67] = "c";
    KeyDown[KeyDown["d"] = 68] = "d";
    KeyDown[KeyDown["e"] = 69] = "e";
    KeyDown[KeyDown["f"] = 70] = "f";
    KeyDown[KeyDown["g"] = 71] = "g";
    KeyDown[KeyDown["h"] = 72] = "h";
    KeyDown[KeyDown["i"] = 73] = "i";
    KeyDown[KeyDown["j"] = 74] = "j";
    KeyDown[KeyDown["k"] = 75] = "k";
    KeyDown[KeyDown["l"] = 76] = "l";
    KeyDown[KeyDown["m"] = 77] = "m";
    KeyDown[KeyDown["n"] = 78] = "n";
    KeyDown[KeyDown["o"] = 79] = "o";
    KeyDown[KeyDown["p"] = 80] = "p";
    KeyDown[KeyDown["q"] = 81] = "q";
    KeyDown[KeyDown["r"] = 82] = "r";
    KeyDown[KeyDown["s"] = 83] = "s";
    KeyDown[KeyDown["t"] = 84] = "t";
    KeyDown[KeyDown["u"] = 85] = "u";
    KeyDown[KeyDown["v"] = 86] = "v";
    KeyDown[KeyDown["w"] = 87] = "w";
    KeyDown[KeyDown["x"] = 88] = "x";
    KeyDown[KeyDown["y"] = 89] = "y";
    KeyDown[KeyDown["z"] = 90] = "z";
    KeyDown[KeyDown["LeftWindowKey"] = 91] = "LeftWindowKey";
    KeyDown[KeyDown["RightWindowKey"] = 92] = "RightWindowKey";
    KeyDown[KeyDown["SelectKey"] = 93] = "SelectKey";
    KeyDown[KeyDown["Numpad0"] = 96] = "Numpad0";
    KeyDown[KeyDown["Numpad1"] = 97] = "Numpad1";
    KeyDown[KeyDown["Numpad2"] = 98] = "Numpad2";
    KeyDown[KeyDown["Numpad3"] = 99] = "Numpad3";
    KeyDown[KeyDown["Numpad4"] = 100] = "Numpad4";
    KeyDown[KeyDown["Numpad5"] = 101] = "Numpad5";
    KeyDown[KeyDown["Numpad6"] = 102] = "Numpad6";
    KeyDown[KeyDown["Numpad7"] = 103] = "Numpad7";
    KeyDown[KeyDown["Numpad8"] = 104] = "Numpad8";
    KeyDown[KeyDown["Numpad9"] = 105] = "Numpad9";
    KeyDown[KeyDown["Multiply"] = 106] = "Multiply";
    KeyDown[KeyDown["NumpadPlus"] = 107] = "NumpadPlus";
    KeyDown[KeyDown["NumpadMinus"] = 109] = "NumpadMinus";
    KeyDown[KeyDown["DecimalPoint"] = 110] = "DecimalPoint";
    KeyDown[KeyDown["Divide"] = 111] = "Divide";
    KeyDown[KeyDown["F1"] = 112] = "F1";
    KeyDown[KeyDown["F2"] = 113] = "F2";
    KeyDown[KeyDown["F3"] = 114] = "F3";
    KeyDown[KeyDown["F4"] = 115] = "F4";
    KeyDown[KeyDown["F5"] = 116] = "F5";
    KeyDown[KeyDown["F6"] = 117] = "F6";
    KeyDown[KeyDown["F7"] = 118] = "F7";
    KeyDown[KeyDown["F8"] = 119] = "F8";
    KeyDown[KeyDown["F9"] = 120] = "F9";
    KeyDown[KeyDown["F10"] = 121] = "F10";
    KeyDown[KeyDown["F11"] = 122] = "F11";
    KeyDown[KeyDown["F12"] = 123] = "F12";
    KeyDown[KeyDown["NumLock"] = 144] = "NumLock";
    KeyDown[KeyDown["ScrollLock"] = 145] = "ScrollLock";
    KeyDown[KeyDown["Semicolon"] = 186] = "Semicolon";
    KeyDown[KeyDown["Equals"] = 187] = "Equals";
    KeyDown[KeyDown["Comma"] = 188] = "Comma";
    KeyDown[KeyDown["LessThan"] = 188] = "LessThan";
    KeyDown[KeyDown["Dash"] = 189] = "Dash";
    KeyDown[KeyDown["Period"] = 190] = "Period";
    KeyDown[KeyDown["GreaterThan"] = 190] = "GreaterThan";
    KeyDown[KeyDown["ForwardSlash"] = 191] = "ForwardSlash";
    KeyDown[KeyDown["QuestionMark"] = 191] = "QuestionMark";
    KeyDown[KeyDown["GraveAccent"] = 192] = "GraveAccent";
    KeyDown[KeyDown["Tilde"] = 192] = "Tilde";
    KeyDown[KeyDown["OpenCurlyBracket"] = 219] = "OpenCurlyBracket";
    KeyDown[KeyDown["OpenSquareBracket"] = 219] = "OpenSquareBracket";
    KeyDown[KeyDown["BackSlash"] = 220] = "BackSlash";
    KeyDown[KeyDown["VerticalPipe"] = 220] = "VerticalPipe";
    KeyDown[KeyDown["CloseCurlyBracket"] = 221] = "CloseCurlyBracket";
    KeyDown[KeyDown["CloseSquareBracket"] = 221] = "CloseSquareBracket";
    KeyDown[KeyDown["Quote"] = 222] = "Quote";
    KeyDown[KeyDown["CommandFF"] = 224] = "CommandFF";
})(KeyDown || (KeyDown = {}));
export var KeyPress;
(function (KeyPress) {
    KeyPress[KeyPress["Backspace"] = 8] = "Backspace";
    KeyPress[KeyPress["Enter"] = 13] = "Enter";
    KeyPress[KeyPress["Spacebar"] = 32] = "Spacebar";
    KeyPress[KeyPress["Hash"] = 35] = "Hash";
    KeyPress[KeyPress["GraveAccent"] = 39] = "GraveAccent";
    KeyPress[KeyPress["DoubleQuote"] = 34] = "DoubleQuote";
    KeyPress[KeyPress["Asterisk"] = 42] = "Asterisk";
    KeyPress[KeyPress["Plus"] = 43] = "Plus";
    KeyPress[KeyPress["Comma"] = 44] = "Comma";
    KeyPress[KeyPress["Minus"] = 45] = "Minus";
    KeyPress[KeyPress["Period"] = 46] = "Period";
    KeyPress[KeyPress["ForwardSlash"] = 47] = "ForwardSlash";
    KeyPress[KeyPress["Zero"] = 48] = "Zero";
    KeyPress[KeyPress["One"] = 49] = "One";
    KeyPress[KeyPress["Two"] = 50] = "Two";
    KeyPress[KeyPress["Three"] = 51] = "Three";
    KeyPress[KeyPress["Four"] = 52] = "Four";
    KeyPress[KeyPress["Five"] = 53] = "Five";
    KeyPress[KeyPress["Six"] = 54] = "Six";
    KeyPress[KeyPress["Seven"] = 55] = "Seven";
    KeyPress[KeyPress["Eight"] = 56] = "Eight";
    KeyPress[KeyPress["Nine"] = 57] = "Nine";
    KeyPress[KeyPress["Colon"] = 58] = "Colon";
    KeyPress[KeyPress["Semicolon"] = 59] = "Semicolon";
    KeyPress[KeyPress["LessThan"] = 60] = "LessThan";
    KeyPress[KeyPress["Equals"] = 61] = "Equals";
    KeyPress[KeyPress["GreaterThan"] = 62] = "GreaterThan";
    KeyPress[KeyPress["QuestionMark"] = 63] = "QuestionMark";
    KeyPress[KeyPress["At"] = 64] = "At";
    KeyPress[KeyPress["OpenSquareBracket"] = 91] = "OpenSquareBracket";
    KeyPress[KeyPress["BackSlash"] = 92] = "BackSlash";
    KeyPress[KeyPress["CloseSquareBracket"] = 93] = "CloseSquareBracket";
    KeyPress[KeyPress["a"] = 97] = "a";
    KeyPress[KeyPress["b"] = 98] = "b";
    KeyPress[KeyPress["c"] = 99] = "c";
    KeyPress[KeyPress["d"] = 100] = "d";
    KeyPress[KeyPress["e"] = 101] = "e";
    KeyPress[KeyPress["f"] = 102] = "f";
    KeyPress[KeyPress["g"] = 103] = "g";
    KeyPress[KeyPress["h"] = 104] = "h";
    KeyPress[KeyPress["i"] = 105] = "i";
    KeyPress[KeyPress["j"] = 106] = "j";
    KeyPress[KeyPress["k"] = 107] = "k";
    KeyPress[KeyPress["l"] = 108] = "l";
    KeyPress[KeyPress["m"] = 109] = "m";
    KeyPress[KeyPress["n"] = 110] = "n";
    KeyPress[KeyPress["o"] = 111] = "o";
    KeyPress[KeyPress["p"] = 112] = "p";
    KeyPress[KeyPress["q"] = 113] = "q";
    KeyPress[KeyPress["r"] = 114] = "r";
    KeyPress[KeyPress["s"] = 115] = "s";
    KeyPress[KeyPress["t"] = 116] = "t";
    KeyPress[KeyPress["u"] = 117] = "u";
    KeyPress[KeyPress["v"] = 118] = "v";
    KeyPress[KeyPress["w"] = 119] = "w";
    KeyPress[KeyPress["x"] = 120] = "x";
    KeyPress[KeyPress["y"] = 121] = "y";
    KeyPress[KeyPress["z"] = 122] = "z";
    KeyPress[KeyPress["OpenCurlyBracket"] = 123] = "OpenCurlyBracket";
    KeyPress[KeyPress["VerticalPipe"] = 124] = "VerticalPipe";
    KeyPress[KeyPress["CloseCurlyBracket"] = 125] = "CloseCurlyBracket";
    KeyPress[KeyPress["Tilde"] = 126] = "Tilde";
})(KeyPress || (KeyPress = {}));
