// @iiif/manifold/dist-commonjs/AnnotationGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AnnotationPoint3D_1 = require("./AnnotationPoint3D");
var AnnotationRect_1 = require("./AnnotationRect");
var AnnotationGroup = /** @class */ (function () {
    function AnnotationGroup(canvasId) {
        this.rects = [];
        this.points3D = [];
        this.canvasId = canvasId;
    }
    AnnotationGroup.prototype.addRect = function (anno) {
        var rect = new AnnotationRect_1.AnnotationRect(anno);
        rect.canvasId = this.canvasId;
        rect.canvasIndex = this.canvasIndex;
        rect.index = this.rects.length;
        this.rects.push(rect);
        // sort ascending
        this.rects.sort(function (a, b) {
            return a.index - b.index;
        });
    };
    AnnotationGroup.prototype.addPoint3D = function (anno) {
        var point = new AnnotationPoint3D_1.AnnotationPoint3D(anno);
        point.canvasId = this.canvasId;
        point.canvasIndex = this.canvasIndex;
        point.index = this.points3D.length;
        this.points3D.push(point);
        // sort ascending
        this.points3D.sort(function (a, b) {
            return a.index - b.index;
        });
    };
    return AnnotationGroup;
}());
exports.AnnotationGroup = AnnotationGroup;
